### ISSUE TITLE
[BUG] numpy implements PEP-3118 incorrectly

### ISSUE THREAD

**Issue by st-pasha** [numpy/numpy#9456]
[PEP 3118](https://www.python.org/dev/peps/pep-3118/) describes Buffers protocol to implement data interchange between different Python/C objects.

I'm implementing such an object, which exposes `PyBuffers` interface for the purpose of interoperability with other libraries. For example, I expect that if `x` is my object, then `numpy.array(x)` would convert it into a numpy array via the Buffers protocol.

What I find however is that `numpy.array()` constructor calls `PyObject_GetBuffer()` on my object with `flags=0x11C` (which is a combination of `PyBUF_INDIRECT | PyBUF_FORMAT`). According to the specification, the `PyBUF_INDIRECT` should be used
> ... when the consumer can handle indirect array referencing implied by these suboffsets.

Unfortunately `numpy` does not support arrays with suboffsets, and therefore either displays incorrect information or causes a seg.fault when trying to read memory without properly following the pointers as required by the spec.

The expected behavior is that `numpy` would not supply the `PyBUF_INDIRECT` flag when calling `PyObject_GetBuffer`, but instead use the `PyBUF_STRIDES` flag which seems to correspond to what `numpy` can actually handle.

**Comment by rkern** [comment]
I don't see where `numpy` does this. Can you identify the specific `PyObj_GetBuffer()` call that uses those arguments? I only see `PyBUF_FORMAT`, `PyBUF_STRIDES`, `PyBUF_ND`, `PyBUF_SIMPLE`, and `PyBUF_ANY_CONTIGUOUS`, none of which include `PyBUF_INDIRECT`.

There is also the `get_buffer_info()` function, but all it does is expose `PyObj_GetBuffer()` to Python converting the string arguments to the flags. It's not called anywhere in `numpy` that I can see.

**Comment by pv** [comment]
Sounds like doing something like `if (flags & PyBUF_INDIRECT)`, which is incorrect since it's a multi-bit mask.

**Comment by pv** [comment]
No, the problem is in `_array_from_buffer_3118`, which first constructs a memoryview object, and then obtains the buffer from that. Python memoryviews do use `PyBUF_FULL`.

The fix is going to require some more work, because it's necessary to deal with the life cycle management of the buffer without using python memoryviews.

**Comment by eric-wieser** [comment]
For now, we can add a check that `view->suboffsets == NULL`, and throw an exception otherwise - which would at least stop the segfaults

**Comment by pv** [comment]
A simple fix would be resurrecting https://github.com/numpy/numpy/blob/c4ee11b61e278/numpy/core/src/multiarray/numpymemoryview.c and using that for the life cycle management.

**Comment by eric-wieser** [comment]
What am I missing here? Why can't we get the buffer ourselves with the right flags, then wrap it in a `memoryview` using `PyMemoryView_FromBuffer`?

**Comment by pv** [comment]
It's necessary to call `PyBuffer_Release` when the buffer is no longer needed. So either we add a new item to the PyArrayObject struct, or manage doing that with a separate Python object.

**Comment by eric-wieser** [comment]
But the `memoryview` destructor does that for us

**Comment by pv** [comment]
Sorry, you asked different question. The problem is that the `PyMemoryView_FromBuffer` does basically memcpy of the `Py_buffer` struct.

However, it is allowed and also the CPython implementation used to store static data in the `Py_buffer` structure itself and *point to it with pointers stored in the same structure*. That cannot be copied with just memcpy, and does not work with `PyMemoryView_FromBuffer`.

I think this is something that may have changed in newer Python 3.x releases.

**Comment by eric-wieser** [comment]
> The CPython implementation used to store static data in the Py_buffer structure itself and point to it with pointers stored in the same structure. 

Store data where? The only place would be `internal`, and that's a pointer, so is fine to refer to a self-referential structure. I suppose that might refer back to the buffer though.

How was `PyMemoryView_FromBuffer` supposed to be used then?

**Comment by pv** [comment]
In earlier python versions the structure had additional fields. These seem to have gone away in recent Python releases.

The first releases of memoryview in Python were buggy in this respect. Indeed, PyMemoryView_FromBuffer could not be used.

**Comment by pv** [comment]
This issue with the implementation in CPython still persists afaik in Python 2.7.

Basically, `PyBuffer_FillInfo` still does stuff like
```
    view->shape = NULL;
    if ((flags & PyBUF_ND) == PyBUF_ND)
        view->shape = &(view->len);
    view->strides = NULL;
    if ((flags & PyBUF_STRIDES) == PyBUF_STRIDES)
        view->strides = &(view->itemsize);
```
and the builtin memoryview object uses the `smalltable` static field.

**Comment by eric-wieser** [comment]
Isn't that fixed [here](https://github.com/python/cpython/blob/2.7/Objects/memoryobject.c#L19-L31)?

**Comment by pv** [comment]
Only if assuming it only CPython does this.

**Comment by njsmith** [comment]
What's wrong with the suggestion of using `memoryview` but then erroring out if it has suboffsets? That seems much simpler than defining a whole new class just for this.

**Comment by eric-wieser** [comment]
[Related python bug](https://bugs.python.org/issue10181) concerning `PyMemoryView_FromBuffer`

**Comment by pv** [comment]
@njsmith: the object may also able to provide a view without suboffsets if requested.

I don't recall what was the outcome from the Cpython discussion --- whether this was ruled as something that only CPython is allowed to do, or if it's free for all. Of course, we could in principle use the *FromBuffer and declare non-complying implementations wrong.

**Comment by njsmith** [comment]
> the object may also able to provide a view without suboffsets if requested.

Is that realistic? Like, either an object's memory layout requires suboffsets or it doesn't, right? And if it does then the only way to switch it is to copy the data, which we can just as well do ourselves?

**Comment by pv** [comment]
I would imagine there are few projects that implement suboffsets at all,
given this is the first report in 7 or so years...

25.7.2017 3.05 "Nathaniel J. Smith" <notifications@github.com> kirjoitti:

**Comment by eric-wieser** [comment]
[Another related bug - issue15821](https://bugs.python.org/issue15821)

**Comment by nwhitehorn** [comment]
`

>> the object may also able to provide a view without suboffsets if requested.
>
> Is that realistic? Like, either an object's memory layout requires suboffsets or it doesn't, right? And if it does then the only way to switch it is to copy the data, which we can just as well do ourselves?`

It is a little realistic. I hit this bug yesterday trying to construct a buffer to a 2D array made of scattered 1D arrays. From GetBuffer(), I can iterate through the ~10K single arrays to construct a contiguous block substantially faster than they can be iterated through in Python in the fallback path. (Right now, this is in fact the *only* interface I can provide because otherwise numpy will segfault if I have code to generate a buffer with suboffsets.)

As an aside, it would of course be nice if numpy grew support for irregularly discontiguous arrays with suboffsets, but that's beyond the immediate scope.

**Comment by abalkin** [comment]
> Unfortunately numpy does not support arrays with suboffsets, ...

This was reported in #5412.
